File: golang/brackets/brackets_test.go
package brackets

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func Test_Empty_Success(t *testing.T) {
	err := BalancedBrackets("")
	require.NoError(t, err)
}

func Test_SimpleBrackets_Success(t *testing.T) {
	err := BalancedBrackets("()")
	require.NoError(t, err)
}

func Test_SimpleBrackets_Fail_RemainingOpen(t *testing.T) {
	err := BalancedBrackets("(()")
	require.ErrorContains(t, err, "opening brackets without closing brackets found: [(]")
}

func Test_SimpleBrackets_Fail_ExtraClosed(t *testing.T) {
	err := BalancedBrackets("(()))")
	require.ErrorContains(t, err, "closing bracket ) with no opening bracket at rune 5")
}

func Test_SimpleBrackets_Fail_Mismatched(t *testing.T) {
	err := BalancedBrackets("((])")
	require.ErrorContains(t, err, "closing bracket ] at rune 3 mismatched with last opening bracket (")
}

func Test_ComplexText_Success(t *testing.T) {
	err := BalancedBrackets("This is a bit of (albeit ridiculous) explanatory text. Don't (forget (to nest).).")
	require.NoError(t, err)
}

func Test_ComplexText_Mismatched(t *testing.T) {
	err := BalancedBrackets("This is a bit of (albeit ridiculous] explanatory text. Don't (forget (to nest).).")
	require.ErrorContains(t, err, "closing bracket ] at rune 36 mismatched with last opening bracket (")
}

File: kotlin/src/test/kotlin/bracketskt/BracketsTest.kt
package bracketskt

import com.geekinasuit.polyglot.example.protos.Something
import com.google.common.truth.Truth.assertThat
import org.junit.Assert.assertThrows
import org.junit.Test

class BracketsTest {

  @Test
  fun testEmptySuccess() { 
    balancedBrackets("")
  }

  @Test
  fun testSimpleBracketsSuccess() {
    balancedBrackets("()")
  }

  @Test
  fun testSimpleBracketsFailRemainingOpen() {
    val err = assertThrows(BracketsNotBalancedException::class.java) { balancedBrackets("(()") }
    assertThat(err).hasMessageThat().contains("opening brackets without closing brackets found: [(]")
  }

  @Test
  fun testSimpleBracketsFailExtraClosed() {
    val err = assertThrows(BracketsNotBalancedException::class.java) { balancedBrackets("(()))") }
    assertThat(err).hasMessageThat().contains("closing bracket ) with no opening bracket at char 5")

  }

  @Test
  fun testSimpleBracketsFailMismatched() {
    val err = assertThrows(BracketsNotBalancedException::class.java) { balancedBrackets("((])") }
    assertThat(err).hasMessageThat().contains("closing bracket ] at char 3 mismatched with last opening bracket (")
  }

  @Test
  fun testComplexTextSuccess() {
    balancedBrackets("This is a bit of (albeit ridiculous) explanatory text. Don't (forget (to nest).).")
  }

  @Test
  fun testComplexMixedTextSuccess() {
    balancedBrackets("This is a bit of (albeit ridiculous) explanatory text. Don't (forget [to {nest}].).")
  }

  @Test
  fun testComplexTextMismatched() {
    val err = assertThrows(BracketsNotBalancedException::class.java) {
      balancedBrackets("This is a bit of (albeit ridiculous] explanatory text. Don't (forget (to nest).).")
    }
    assertThat(err).hasMessageThat().contains("closing bracket ] at char 36 mismatched with last opening bracket (")
  }
  @Test
  fun testFoo() {
    val actual: Something = foo()
    assertThat(actual.id).isEqualTo(2)
    assertThat(actual.name).isEqualTo("foo")
    assertThat(actual.labelsList).containsExactly("a", "b").inOrder()
  }
}
